File: backend/api/contact_messages.py
"""
Contact Messages API
Endpoints for contact message management
"""

from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.ext.asyncio import AsyncSession
from typing import Optional
from uuid import UUID
import math

from core.db import get_db
from core.dependencies import get_current_user, require_admin
from schemas.contact_message import (
    ContactMessageCreate,
    ContactMessageUpdate,
    ContactMessageResponse,
    ContactMessageListResponse,
    MessageStatus,
    MessagePriority
)
from core.utils.response import Response
from core.errors import APIException
from services.contact_message import ContactMessageService
from core.logging import get_logger

logger = get_logger(__name__)
router = APIRouter(prefix="/contact-messages", tags=["Contact Messages"])


@router.post("")
async def create_contact_message(
    message_data: ContactMessageCreate,
    db: AsyncSession = Depends(get_db)
):
    """
    Create a new contact message (public endpoint)
    """
    try:
        message = await ContactMessageService.create_message(db, message_data)
        return Response.success(
            data={
                "id": str(message.id),
                "name": message.name,
                "email": message.email,
                "subject": message.subject,
                "message": message.message,
                "status": message.status.value,
                "priority": message.priority.value,
                "admin_notes": message.admin_notes,
                "assigned_to": str(message.assigned_to) if message.assigned_to else None,
                "created_at": message.created_at.isoformat() if hasattr(message.created_at, 'isoformat') else str(message.created_at),
                "updated_at": message.updated_at.isoformat() if hasattr(message.updated_at, 'isoformat') else str(message.updated_at),
                "resolved_at": message.resolved_at.isoformat() if message.resolved_at and hasattr(message.resolved_at, 'isoformat') else str(message.resolved_at) if message.resolved_at else None
            },
            message="Contact message created successfully",
            status_code=status.HTTP_201_CREATED
        )
    except Exception as e:
        logger.error(f"Error creating contact message: {str(e)}")
        raise APIException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            message="Failed to create contact message"
        )


@router.get("")
async def get_all_contact_messages(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    status_filter: Optional[MessageStatus] = Query(None, alias="status"),
    priority: Optional[MessagePriority] = None,
    search: Optional[str] = None,
    db: AsyncSession = Depends(get_db),
    current_user = Depends(require_admin)
):
    """
    Get all contact messages with pagination and filters (admin only)
    """
    try:
        messages, total = await ContactMessageService.get_all_messages(
            db=db,
            page=page,
            page_size=page_size,
            status=status_filter,
            priority=priority,
            search=search
        )
        
        total_pages = math.ceil(total / page_size) if total > 0 else 0
        
        return Response.success(
            data={
                "messages": [
                    {
                        "id": str(msg.id),
                        "name": msg.name,
                        "email": msg.email,
                        "subject": msg.subject,
                        "message": msg.message,
                        "status": msg.status.value,
                        "priority": msg.priority.value,
                        "admin_notes": msg.admin_notes,
                        "assigned_to": str(msg.assigned_to) if msg.assigned_to else None,
                        "created_at": msg.created_at.isoformat() if hasattr(msg.created_at, 'isoformat') else str(msg.created_at),
                        "updated_at": msg.updated_at.isoformat() if hasattr(msg.updated_at, 'isoformat') else str(msg.updated_at),
                        "resolved_at": msg.resolved_at.isoformat() if msg.resolved_at and hasattr(msg.resolved_at, 'isoformat') else str(msg.resolved_at) if msg.resolved_at else None
                    }
                    for msg in messages
                ]
            },
            message="Contact messages retrieved successfully",
            pagination={
                "page": page,
                "page_size": page_size,
                "total": total,
                "total_pages": total_pages
            }
        )
    except Exception as e:
        logger.error(f"Error fetching contact messages: {str(e)}")
        raise APIException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            message="Failed to fetch contact messages"
        )


@router.get("/stats")
async def get_contact_message_stats(
    db: AsyncSession = Depends(get_db),
    current_user = Depends(require_admin)
):
    """
    Get contact message statistics (admin only)
    """
    try:
        stats = await ContactMessageService.get_message_stats(db)
        return Response.success(
            data=stats,
            message="Contact message statistics retrieved successfully"
        )
    except Exception as e:
        logger.error(f"Error fetching contact message stats: {str(e)}")
        raise APIException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            message="Failed to fetch statistics"
        )


@router.get("/{message_id}")
async def get_contact_message(
    message_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user = Depends(require_admin)
):
    """
    Get a specific contact message by ID (admin only)
    """
    message = await ContactMessageService.get_message_by_id(db, message_id)
    
    if not message:
        raise APIException(
            status_code=status.HTTP_404_NOT_FOUND,
            message="Contact message not found"
        )
    
    return Response.success(
        data={
            "id": str(message.id),
            "name": message.name,
            "email": message.email,
            "subject": message.subject,
            "message": message.message,
            "status": message.status.value,
            "priority": message.priority.value,
            "admin_notes": message.admin_notes,
            "assigned_to": str(message.assigned_to) if message.assigned_to else None,
            "created_at": message.created_at.isoformat() if hasattr(message.created_at, 'isoformat') else str(message.created_at),
            "updated_at": message.updated_at.isoformat() if hasattr(message.updated_at, 'isoformat') else str(message.updated_at),
            "resolved_at": message.resolved_at.isoformat() if message.resolved_at and hasattr(message.resolved_at, 'isoformat') else str(message.resolved_at) if message.resolved_at else None
        },
        message="Contact message retrieved successfully"
    )


@router.patch("/{message_id}")
async def update_contact_message(
    message_id: UUID,
    update_data: ContactMessageUpdate,
    db: AsyncSession = Depends(get_db),
    current_user = Depends(require_admin)
):
    """
    Update a contact message (admin only)
    """
    try:
        message = await ContactMessageService.update_message(db, message_id, update_data)
        
        if not message:
            raise APIException(
                status_code=status.HTTP_404_NOT_FOUND,
                message="Contact message not found"
            )
        
        return Response.success(
            data={
                "id": str(message.id),
                "name": message.name,
                "email": message.email,
                "subject": message.subject,
                "message": message.message,
                "status": message.status.value,
                "priority": message.priority.value,
                "admin_notes": message.admin_notes,
                "assigned_to": str(message.assigned_to) if message.assigned_to else None,
                "created_at": message.created_at.isoformat() if hasattr(message.created_at, 'isoformat') else str(message.created_at),
                "updated_at": message.updated_at.isoformat() if hasattr(message.updated_at, 'isoformat') else str(message.updated_at),
                "resolved_at": message.resolved_at.isoformat() if message.resolved_at and hasattr(message.resolved_at, 'isoformat') else str(message.resolved_at) if message.resolved_at else None
            },
            message="Contact message updated successfully"
        )
    except APIException:
        raise
    except Exception as e:
        logger.error(f"Error updating contact message: {str(e)}")
        raise APIException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            message="Failed to update contact message"
        )


@router.delete("/{message_id}")
async def delete_contact_message(
    message_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user = Depends(require_admin)
):
    """
    Delete a contact message (admin only)
    """
    try:
        success = await ContactMessageService.delete_message(db, message_id)
        
        if not success:
            raise APIException(
                status_code=status.HTTP_404_NOT_FOUND,
                message="Contact message not found"
            )
        
        return Response.success(
            message="Contact message deleted successfully"
        )
    except APIException:
        raise
    except Exception as e:
        logger.error(f"Error deleting contact message: {str(e)}")
        raise APIException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            message="Failed to delete contact message"
        )

File: frontend/src/api/contact-messages.ts
/**
 * Contact Messages API Client
 */

import apiClient from './client';

export interface ContactMessageCreate {
  name: string;
  email: string;
  subject: string;
  message: string;
}

export interface ContactMessage {
  id: string;
  name: string;
  email: string;
  subject: string;
  message: string;
  status: 'new' | 'in_progress' | 'resolved' | 'closed';
  priority: 'low' | 'medium' | 'high' | 'urgent';
  admin_notes?: string;
  assigned_to?: string;
  created_at: string;
  updated_at: string;
  resolved_at?: string;
}

export interface ContactMessageUpdate {
  status?: 'new' | 'in_progress' | 'resolved' | 'closed';
  priority?: 'low' | 'medium' | 'high' | 'urgent';
  admin_notes?: string;
  assigned_to?: string;
}

export interface ContactMessageListResponse {
  messages: ContactMessage[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export const ContactMessagesAPI = {
  /**
   * Create a new contact message (public)
   */
  create: async (data: ContactMessageCreate): Promise<ContactMessage> => {
    const response = await apiClient.post('/contact-messages', data);
    return response.data.data;
  },

  /**
   * Get all contact messages (admin only)
   */
  getAll: async (params?: {
    page?: number;
    page_size?: number;
    status?: string;
    priority?: string;
    search?: string;
  }): Promise<ContactMessageListResponse> => {
    const response = await apiClient.get('/contact-messages', { params });
    
    // The backend returns: { success, data: { messages: [...] }, pagination: {...} }
    // So we need response.data.data.messages
    const responseData = response.data;
    
    // Check if we have the expected structure
    if (responseData?.data?.messages && Array.isArray(responseData.data.messages)) {
      return {
        messages: responseData.data.messages,
        total: responseData.pagination?.total || 0,
        page: responseData.pagination?.page || params?.page || 1,
        page_size: responseData.pagination?.page_size || params?.page_size || 20,
        total_pages: responseData.pagination?.total_pages || 0
      };
    }
    
    // Fallback: if messages are directly in data (shouldn't happen but handle it)
    if (responseData?.messages && Array.isArray(responseData.messages)) {
      console.warn('Unexpected response structure - messages directly in data');
      return {
        messages: responseData.messages,
        total: responseData.total || 0,
        page: responseData.page || params?.page || 1,
        page_size: responseData.page_size || params?.page_size || 20,
        total_pages: responseData.total_pages || 0
      };
    }
    
    // No valid data found
    console.error('Invalid response structure:', responseData);
    return {
      messages: [],
      total: 0,
      page: params?.page || 1,
      page_size: params?.page_size || 20,
      total_pages: 0
    };
  },

  /**
   * Get contact message statistics (admin only)
   */
  getStats: async (): Promise<{
    total: number;
    new: number;
    in_progress: number;
    resolved: number;
  }> => {
    try {
      const response = await apiClient.get('/contact-messages/stats');
      return response.data.data || { total: 0, new: 0, in_progress: 0, resolved: 0 };
    } catch (error) {
      console.error('Error fetching stats:', error);
      return { total: 0, new: 0, in_progress: 0, resolved: 0 };
    }
  },

  /**
   * Get a specific contact message (admin only)
   */
  getById: async (id: string): Promise<ContactMessage> => {
    const response = await apiClient.get(`/contact-messages/${id}`);
    return response.data.data;
  },

  /**
   * Update a contact message (admin only)
   */
  update: async (id: string, data: ContactMessageUpdate): Promise<ContactMessage> => {
    const response = await apiClient.patch(`/contact-messages/${id}`, data);
    return response.data.data;
  },

  /**
   * Delete a contact message (admin only)
   */
  delete: async (id: string): Promise<void> => {
    await apiClient.delete(`/contact-messages/${id}`);
  },
};

export default ContactMessagesAPI;
